### ISSUE TITLE
EB not activated in TRIP

### ISSUE THREAD

**Issue by jokaICS** [openETCS/modeling#928]
When the EVC receives an UES (Msg16), the mode is switched to TRIP (indicated by DMI and Msg146), but the emergency brakes are not applied.

**Comment by BerndHekele** [comment]
@stefan-karg @T12z 
Hi Thorsten,
I can comform this statement from earlier tests.
In my understanding without having had a recent look at the code for this issue I understand this is likely to be  wrong in the SDM function. Can you make a quick statement on this issue?

**Comment by T12z** [comment]
If you relate to EB from M&L, this is quite possible, I never really tested it. Awkward I noticed this myself though never had a deeper look at my hot fix. I will also verify trip from passing EoA, though I have seen this working before.

I am currently waiting on a real-life Signalstörung at the moment ;-) 

---- Bernd Hekele schrieb ----

**Comment by T12z** [comment]
So, I had a first look,
can you please specify, how to reach that TRIP and and what should happen (i.e. does the trip come frome M&L?)

On a side note on tripping due to EoA overpassing:
In SDM's code, I noticed that there are *two* trip operators. One seems related to a wish of M&L and creates `eoaOverpassed`. A different one calculates an SDM-internal "trip" that triggers the EB.
Now I actually cannot find a spec point that this internal trip should trigger the EB directly as it does. Also these two operators have slightly different calculations. The second one relates correctly to my spec part, and I have no idea what the first one relates to :confused: 

I only noticed, the current implementation should trigger the EB, but cannot release - I fixed that. I will run a test next.

**Comment by BerndHekele** [comment]
The trip is initiated by M&L and M&L also triggers EB/SB.

**Comment by T12z** [comment]
How? If I use SimCtrl and press [ UES ] nothing happens.

For eaxmple, when changing from NTC to L2 M&L toggled the SB for one cycle. So what does that boolean request actually mean?? So far I interprete those two bools coming from M&L as level sensitive. So switching back to request == false also means to release the brakes again. Is this actually an interface issue?

**Comment by jokaICS** [comment]
@T12z the UES button in the current SimCtrl on gitHub is currently still a dummy :( I'll push an updated version with a working UES button soon (~1h)

**Comment by T12z** [comment]
I pushed a correction, but that is actually rather related to releasing after a trip and tripping on EoA did not occur in some situations (eg when SvL-target-limitation sneakes in front of EoA).
Again: I handle the inputs from M&L as level-sensitive - once gone I release the EB.

**Comment by MariellePetitDoche** [comment]
There is plenty of condition to go in Trip mode and simultaneously request EB.
I think the simple one is to have a Msg 16. Then the train shall be at standstill to left Trip mode (for example via revocation).

For the SB toggled during transition from NTC and L2, it can be link to a late acknowledgement of the driver.

**Comment by jokaICS** [comment]
@T12z I've pushed a new version of SimCtrl/libenvsim which allows triggering of a MSG16 via the button on the Commands tab.
Note: this dynamic triggering currently only works when running the ROOT_Scripted operator.

**Comment by T12z** [comment]
@MariellePetitDoche, with the two signals (requestEB, requestSB), is it correct to release the related brake on the true -> false transition of the signal? There is no other revokation signal

**Comment by MariellePetitDoche** [comment]
@T12z I do not know the signals you speak about.
1. the train is in Trip mode
2. the driver ack the trip and the train is at standstill
3. then, for level 1,2 or 3, the train passes in post-trip mode
4. then, in level 2 or 3, RBC shall send Msg 18

**Comment by T12z** [comment]
Sorry @MariellePetitDoche , to clarify the signals I talk about: ManageLevelAndMode outputs `EB_Request` ("requestEB") and `Serive_Brake_Command` ("requestSB"). So far I am only worried about SDM and SDM is **not** aware of modes and procedures.
Anyways, are those signals held *true* until PT is entered, or is it just triggered for one cycle?

I think we need to postpone this discussion until next week and gather for a few minutes about this in MUC. Sorry, I wont be able to modify the/my model over the weekend / on the road until then.

**Comment by BerndHekele** [comment]
I analysed the problem with the current evc model.
The EB is not automatically triggered by the trip procedure of Mode and Level.
I added an trigger to transport this information.
@MariellePetitDoche : Do you see this as a function of Mode Management?

**Comment by MariellePetitDoche** [comment]
We have a discussion yesterday with @T12z on this issue to be agree on the meaning on the signals.
I wait his analyse of the issue after this discussion.
I'am sure that EB is not trigger in Trip mode by M&L Management function. Indeed in the current architecture of the EVC, it is not clear which function manage the release of EB (which can be request by several function).

For a quick correction, this can be added in ML management (in the case of trip mode), but a general analyse how are managed EB and SB by EVC will be beter in my point of view.

**Comment by BerndHekele** [comment]
I don't actually need a correction from M&L if it is not part of the requirement to Modes.
I will add this workaround to the EVC model for the time being.

cu
Bernd

**Comment by BerndHekele** [comment]
Even though the discussion is not finalized we can now proceed with the User Story 15.

**Comment by MariellePetitDoche** [comment]
I have added the EB request for mode TR, SF and NP in the moes management model.

**Comment by T12z** [comment]
I have been a bit late answering on this.
Anyways, as discussed with Marielle last week, we have a way to enter trip-braking via a signal from M&L to SDM. Once the signal falls, SDM may release the brake if allowable. But this approach is an intermediate solution. There must be a new architecture with a well defined brake concept. SDM is not defined for all modes, so it should only be enabled when defined. There is a module that is not implemented yet, that is Movement Monitoring (essentially throwing the brake when there should be no movement, or no further movement) This would would need some larger analyses how the current model architecture needs change

**Comment by T12z** [comment]
Being the assignee,
and since there is nothing left to fix in SDM concerning this and a work around has been provided, I will close. Thanks all.

**Comment by BerndHekele** [comment]
I'll take this as tzhe example for the WP3-Demo!!!
